app/src/main/java/com/onewelcome/showcaseapp/feature/userregistration/twostepregistration/TwoStepRegistrationViewModel.kt

```diff
@@ -0,0 +1,220 @@
+  private val createPinRequestHandler: CreatePinRequestHandler,
+) : ViewModel() {
+
+  var uiState by mutableStateOf(State())
+    private set
+
+  private val _navigationEvents = Channel<NavigationEvent>(Channel.BUFFERED)
+  val navigationEvents = _navigationEvents.receiveAsFlow()
+
+  init {
+    viewModelScope.launch {
+      isSdkInitializedUseCase.execute().let { uiState = uiState.copy(isSdkInitialized = it) }
+      updateIdentityProviders()
+      updateSelectedIdentityProvider()
+      updateUserProfiles()
+      updateCancellationButton()
+    }
+  }
+
+  private fun enterOptionalData() {
+    viewModelScope.launch {
+      _navigationEvents.send(NavigationEvent.ToTwoStepInputScreen)
+    }
+  }
+
+  fun onEvent(event: UiEvent) {
+    when (event) {
+      is UiEvent.StartTwoStepRegistration -> startRegistration()
+      is UiEvent.UpdateSelectedIdentityProvider -> uiState = uiState.copy(selectedIdentityProvider = event.identityProvider)
+      is UiEvent.UpdateSelectedScopes -> uiState = uiState.copy(selectedScopes = event.scopes)
+      is UiEvent.CancelRegistration -> cancelRegistration()
+      is UiEvent.SetStatelessRegistration -> uiState = uiState.copy(isStatelessRegistration = event.isStateless)
+      is UiEvent.SubmitResponseCode -> submitResponseCode(event.responseCode)
+      is UiEvent.UpdateResponseCode -> uiState = uiState.copy(responseCode = event.code)
+      is UiEvent.SubmitOptionalData -> subnitOptionalData(event.optionalData)
+      is UiEvent.EnterTwoStepInput -> enterOptionalData()
+    }
+  }
+
+  private fun subnitOptionalData(optionalData: String) {
+    twoStepRegistrationRequestHandler.optionalData = optionalData
+    onEvent(UiEvent.StartTwoStepRegistration)
+  }
+
+
+  private fun updateIdentityProviders() {
+    getTwoStepIdentityProvidersUseCase.execute()
+      .onSuccess { uiState = uiState.copy(identityProviders = it) }
+      .onFailure { uiState = uiState.copy(identityProviders = emptySet()) }
+  }
+
+  private suspend fun updateUserProfiles() {
+    getUserProfilesUseCase.execute()
+      .onSuccess { uiState = uiState.copy(userProfileIds = it.map { it.profileId }.toList()) }
+      .onFailure { uiState = uiState.copy(userProfileIds = emptyList()) }
+  }
+
+  private fun updateSelectedIdentityProvider() {
+    val identityProviders = uiState.identityProviders
+    if (identityProviders.isNotEmpty()) {
+      uiState = uiState.copy(selectedIdentityProvider = identityProviders.first())
+    }
+  }
+
+  private fun updateCancellationButton() {
+    val isRegistrationInProgress = createPinRequestHandler.isPinCreationInProgress()
+    uiState = uiState.copy(isRegistrationCancellationEnabled = isRegistrationInProgress)
+  }
+
+  private fun cancelRegistration() {
+    viewModelScope.launch {
+      twoStepRegistrationRequestHandler.cancelRegistration()
+      createPinRequestHandler.cancelPinCreation()
+      uiState = uiState.copy(
```

Comment:
Do not update state here manually. `userRegistrationUseCase.register` should return `handleFailure` and update state properly.